src/v/security/audit/schemas/utils.h

```diff
@@ -10,19 +10,146 @@
+    if constexpr (std::is_same_v<T, model::topic>) {
+        return audit_resource_type::topic;
+    } else if constexpr (std::is_same_v<T, kafka::group_id>) {
+        return audit_resource_type::group;
+    } else if constexpr (std::is_same_v<T, security::acl_cluster_name>) {
+        return audit_resource_type::cluster;
+    } else if constexpr (std::is_same_v<T, kafka::transactional_id>) {
+        return audit_resource_type::transactional_id;
+    } else if constexpr (std::is_same_v<T, security::acl_binding>) {
+        return audit_resource_type::acl_binding;
+    } else if constexpr (std::is_same_v<T, security::acl_binding_filter>) {
+        return audit_resource_type::acl_binding_filter;
+    } else {
+        static_assert(utils::unsupported_type<T>::value, "Unsupported type");
+    }
+}
+
+resource_detail transform_to_resource_detail(auto&& v) {
+    using v_t = std::remove_cvref_t<decltype(v)>;
+    const auto get_resource_name = [](auto&& v) {
+        if constexpr (std::is_same_v<v_t, security::acl_binding>) {
+            return "create acl";
+        } else if constexpr (std::
+                               is_same_v<v_t, security::acl_binding_filter>) {
+            return "delete acl";
+        } else {
+            return v();
+        }
+    };
+
+    const auto binding_to_data =
+      [](const security::acl_binding& binding) -> acl_binding_detail {
+        return {
+          .resource_type = fmt::format("{}", binding.pattern().resource()),
+          .resource_name = binding.pattern().name(),
+          .pattern_type = fmt::format("{}", binding.pattern().pattern()),
+          .acl_principal = fmt::format("{}", binding.entry().principal()),
+          .acl_host = fmt::format("{}", binding.entry().host()),
+          .acl_operation = fmt::format("{}", binding.entry().operation()),
+          .acl_permission = fmt::format("{}", binding.entry().permission())};
+    };
+
+    const auto binding_filter_to_data =
+      [](const security::acl_binding_filter& binding) -> acl_binding_detail {
+        const auto print_or_nothing =
+          [](const auto& val) -> std::optional<ss::sstring> {
+            if (val.has_value()) {
+                return fmt::format("{}", val.value());
+            } else {
+                return std::nullopt;
+            }
+        };
+        return {
+          .resource_type = print_or_nothing(binding.pattern().resource()),
+          .resource_name = print_or_nothing(binding.pattern().name()),
+          .pattern_type = print_or_nothing(binding.pattern().pattern()),
+          .acl_principal = print_or_nothing(binding.entry().principal()),
+          .acl_host = print_or_nothing(binding.entry().host()),
+          .acl_operation = print_or_nothing(binding.entry().operation()),
+          .acl_permission = print_or_nothing(binding.entry().permission())};
+    };
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-lambda-capture"
+    const auto get_resource_data =
```

Comment:
you only call this once, so you can probably get rid of the lambda here entirely, just write a constexpr if-expression to fill in a string then you don't need the pragma.

alternatively, you might be able to use a lambda template parameters instead of captures and avoid the issue.